protocols/virtualfund/virtualfund.go

```diff
@@ -567,7 +567,13 @@ func (c *Connection) expectedProposal() consensus_channel.Proposal {
 		leftAmount = val
 		break
 	}
-	proposal := consensus_channel.NewAddProposal(c.Channel.Id, 0, g, leftAmount)
+
+	// this awkward step is required because Channel.SignNextProposal accepts as input the
+	// proposal expected in the queue, and validates that the next proposal in the queue
+	// matches exactly. I am not sure if validating the turn number is critical!
+	turnNum := c.Channel.ConsensusTurnNum() + 1
+
+	proposal := consensus_channel.NewAddProposal(c.Channel.Id, turnNum, g, leftAmount)
 
 	return proposal
```

Comment:
I don't find this step all that awkward! The expected turnNum should be one greater than the current turnNum.

Consider:
```suggestion
	return consensus_channel.NewAddProposal(
		c.Channel.Id,
		c.Channel.ConsensusTurnNum()+1, // turnNum must increment with every proposal
		g,
		leftAmount,
	)
```

Reply:
What I find awkward is that:

The ConsensusChannel only cares about sequential turn numbers to be able to make certain safety guarantees ("this action does not lose funds", because it's incorporating a single change `X`, and that change is explicitly approved by someone who's validating `X`)

The caller here (a VFO) doesn't really care what turn number causes a guarantee to be added. But it HAS to care, because the ConsensusChannel expects to receive the correct turn number.

So, the caller has to:
- get the consensus turn number from the ConsensusChannel
- immediately ping that turn number (plus 1) back to the ConsensusChannel

We could avoid that by changing the function signature of SignNextProposal. This function could say: "sign the next proposal, if it's proposing change `X`" or even just "sign the next proposal" (assuming that the caller knows what the next proposal is)

IMO, the least-awkward thing would be this:

1. Leader sends a signed proposal `sp` to follower
2. follower's VFO receives it, and checks "Is this proposing the correct change?"
3. follower's VFO calls `ConsensusChannel.ReceiveProposal(sp)`, which validates their signature on `sp` and stores it
4. follower's VFO immediately calls `ConsensusChannel.SignNextProposal(sk)`, which signs the stored `sp` and updates the consensus state

(3 & 4 could even be combined into something like `ConsensusChannel.ValidateAndSignProposal(sp, sk)`. In this way, the ConsensusChannel never has a proposal queue.)

3 + 4 is hard to do right now, because events are received in `Update`, but actions are not taken until `Crank`. That's a nice abstraction too, but it does have this downside.

Reply:
I'm still not convinced this is all that awkward! The handling of `turnNum` is encapsulated inside the `Connection` struct, which is essentially an adapter between VFO and CC.

Existing usage of this function is:

```
ledger := c.Channel
signedProposal, err := ledger.SignNextProposal(c.expectedProposal(), *sk)
```

and

```
ledger := connection.Channel // todo: #420 deprecate - replace with LeaderChannel.Propose workflow

if !ledger.IsLeader() {
	return protocols.SideEffects{}, errors.New("only the proposer can propose a ledger update")
}
sideEffects := protocols.SideEffects{}

signedProposal, err := ledger.Propose(connection.expectedProposal(), *sk)
```

each of which could be further encapsulated - IE, the turnNum detail will never leak into VFO methods.

In any case, this can be punted as potential future refactor work.

Reply:
Yes, I think b9cc49fc20c60a14b2ad8f3059faa5ec36b11b9f is a nice change anyway.